### PR TITLE
[FLINK-5289] Meaningful exception when using value state on non-keyed…

### PR DESCRIPTION
**Pull request by StefanRRichter**
This PR fixes [FLINK-5289] and introduces a meaningful exception when registering a keyed state on non-keyed stream.

### CODE REVIEW

**Comment by StefanRRichter** [comment]
CC @twalthr

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamingRuntimeContext.java [hunk 1]
```diff
@@ -106,17 +109,45 @@ public boolean hasBroadcastVariable(String name) {
...
 	@Override
 	public <T> ReducingState<T> getReducingState(ReducingStateDescriptor<T> stateProperties) {
-		return operator.getKeyedStateStore().getReducingState(stateProperties);
+		KeyedStateStore keyedStateStore = checkPreconditionsAndGetKeyedStateStore(stateProperties);
+
+		try {
+			stateProperties.initializeSerializerUnlessSet(getExecutionConfig());
+			return keyedStateStore.getReducingState(stateProperties);
+		} catch (Exception e) {
+			throw new RuntimeException("Error while getting state", e);
```
Comment (StephanEwen) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamingRuntimeContext.java]: Do we need to re-wrap the exceptions here? Can we simply let the original exception bubble up?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamingRuntimeContext.java [hunk 2]
```diff
@@ -106,17 +109,45 @@ public boolean hasBroadcastVariable(String name) {
...
+		}
+	}
+
+	private KeyedStateStore checkPreconditionsAndGetKeyedStateStore(StateDescriptor<?, ?> stateDescriptor) {
+		Preconditions.checkNotNull(stateDescriptor, "The state properties must not be null");
+		KeyedStateStore keyedStateStore = operator.getKeyedStateStore();
+		Preconditions.checkNotNull(keyedStateStore, "Keyed state store is null. This can only be called after a keyBy.");
```
Comment (StephanEwen) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamingRuntimeContext.java]: How about changing this message to
`Keyed state can only be used on a 'keyed stream', i.e., after a 'keyBy()' operation.`

**Comment by StefanRRichter** [comment]
Updated w.r.t. comments.

**Comment by StephanEwen** [comment]
+1 looks good. Will put this into my merge pipeline

**Pull request closed without merging**